### PR TITLE
feat: Add JsonPrimaryKey to support customizing the primary key

### PR DESCRIPTION
**Pull request by kyle5794**
### Contributor checklist

- [x] My commit messages follow the [Conventional Commit Message Format](https://gist.github.com/stephenparish/9941e89d80e2bc58a153#format-of-the-commit-message)
      For example: `fix: Multiply by appropriate coefficient`, or
      `feat(Calculator): Correctly preserve history`
      Any explanation or long form information in your commit message should be
      in a separate paragraph, separated by a blank line from the primary message
- [ ] Bug fixes include regression tests
- [x] Features include unit/acceptance tests

### CODE REVIEW

File: lib/ash_json_api/resource/transformers/require_primary_key.ex [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+    or includes primary_key section from json_api if it has a composite key
+  "
+
+  use Ash.Dsl.Transformer
+  alias Ash.Dsl.Transformer
+
+  def transform(_resource, dsl) do
+    case Transformer.get_option(dsl, [:json_api, :primary_key], :keys) do
+      nil ->
+        dsl
```
Comment (zachdaniel) [inline lib/ash_json_api/resource/transformers/require_primary_key.ex:12]: You can use `Ash.Resource.primary_key/2` here, if you add a clause to ensure that this check goes after the check that validates/caches primary keys (probably a good idea anyway, so they get a sensible error first):

```elixir
# At the bottom of the module
def after(Ash.Resource.Transformers.CachePrimaryKey), do: true
def after(_), do: false
```

Comment (zachdaniel) [reply lib/ash_json_api/resource/transformers/require_primary_key.ex:12]: Then you can say 
```
case Ash.Resource.primary_key(resource) do
  [_] -> {:ok, dsl}
  _ -> ...
end
```

File: lib/ash_json_api/resource/transformers/require_primary_key.ex [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+        |> Transformer.get_entities([:attributes])
+        |> Enum.filter(& &1.primary_key?)
+        |> Enum.map(& &1.name)
+        |> case do
+          [_only_one_primary_key] -> {:ok, dsl}
+          _ -> raise "AshJsonApi requires primary key when a resource has a composite key"
```
Comment (zachdaniel) [inline lib/ash_json_api/resource/transformers/require_primary_key.ex:18]: Lets use the newer DslError

Here is an example:
```elixir
      raise Ash.Error.Dsl.DslError,
        module: __MODULE__,
        path: [:relationships, relationship.name],
        message:
          "Relationship `#{relationship.name}` expects source field `#{relationship.source_field}` to be defined"
```
I know this is based off of my code from the other transformer, so its my fault :D

**Comment by zachdaniel** [comment]
Had a few comments, but overall this looks great! Awesome work.

As for testing the DSL validations, you can actually run compile time code in the tests, e.g

```
test "foo" do
  assert_raise Ash.Error.Dsl.DslError, ~r/match/, fn -> 
    defmodule BadResource do
       .....
    end
  end
end
```

File: lib/ash_json_api/resource/transformers/require_primary_key.ex (hunk 2 above)
Comment (kyle5794) [reply lib/ash_json_api/resource/transformers/require_primary_key.ex]: Added Ash.Error.Dsl.DslError

File: lib/ash_json_api/resource/transformers/require_primary_key.ex (hunk 1 above)
Comment (kyle5794) [reply lib/ash_json_api/resource/transformers/require_primary_key.ex]: Added `&after?/1`

**Comment by zachdaniel** [comment]
Great work!

🚀 Thank you for your contribution! 🚀

**Pull request merged**